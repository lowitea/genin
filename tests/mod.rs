use std::{fs::File, io::Write, process::Command};

fn create_file(path: &str) {
    File::create(path).unwrap();
}

fn create_test_dir(path: &str) {
    std::fs::create_dir_all(path).expect("failed to create dir");
}

fn delete_test_dir(path: &str) {
    let _ = std::fs::remove_dir_all(path);
}

fn cleanup_test_dir(path: &str) {
    delete_test_dir(path);
    create_test_dir(path);
}

#[test]
fn warning_message_on_init_output() {
    cleanup_test_dir("tests/.tmp1");
    create_file("tests/.tmp1/cluster.genin.yml");

    let output = Command::new(format!(
        "{}/target/debug/genin",
        std::env::var("CARGO_MANIFEST_DIR").unwrap()
    ))
    .arg("init")
    .arg("-q")
    .current_dir("tests/.tmp1")
    .output()
    .expect("Failed to execute command");

    let mut result = Vec::new();
    result.write_all(&output.stdout).unwrap();
    result.write_all(&output.stderr).unwrap();

    assert_eq!(
        result,
        b"WARN: the target file cluster.genin.yml already exists so the new file will \
            be saved with name cluster.genin.copy.yml\n"
    );
}

#[test]
fn warning_message_on_build_output() {
    cleanup_test_dir("tests/.tmp2");
    create_file("tests/.tmp2/inventory.yml");

    Command::new(format!(
        "{}/target/debug/genin",
        std::env::var("CARGO_MANIFEST_DIR").unwrap()
    ))
    .arg("init")
    .arg("-q")
    .current_dir("tests/.tmp2")
    .output()
    .expect("Failed to execute command");

    let output = Command::new(format!(
        "{}/target/debug/genin",
        std::env::var("CARGO_MANIFEST_DIR").unwrap()
    ))
    .arg("build")
    .arg("--source")
    .arg("cluster.genin.yml")
    .arg("-q")
    .current_dir("tests/.tmp2")
    .output()
    .expect("Failed to execute command");

    let mut result = Vec::new();
    result.write_all(&output.stdout).unwrap();
    result.write_all(&output.stderr).unwrap();

    println!("{}", String::from_utf8(result.clone()).unwrap());

    assert_eq!(
        result,
        b"WARN: the target file inventory.yml already exists so the new file will be saved with name inventory.copy.yml\n"
    );
}
