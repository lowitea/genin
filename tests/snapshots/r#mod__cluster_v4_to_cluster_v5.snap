---
source: tests/mod.rs
expression: cluster_v4_to_cluster_v5
---
+-------------+-------------+
|          cluster          |
+-------------+-------------+
|       datacenter-1        |
+-------------+-------------+
|  server-1   |  server-2   |
+-------------+-------------+
|  router-1   |  router-2   |
|  8081/3031  |  8081/3031  |
+-------------+-------------+
| storage-1-1 | storage-1-2 |
| 8082/3032   | 8082/3032   |
+-------------+-------------+
| storage-2-1 | storage-2-2 |
| 8083/3033   | 8083/3033   |
+-------------+-------------+
| storage-3-1 | storage-3-2 |
| 8084/3034   | 8084/3034   |
+-------------+-------------+
| storage-4-1 | storage-4-2 |
| 8085/3035   | 8085/3035   |
+-------------+-------------+

---
all:
  vars:
    ansible_user: vagrant
    ansible_password: vagrant
    cartridge_app_name: new-app
    cartridge_cluster_cookie: myapp-cookie
    cartridge_package_path: /tmp/myapp.rpm
    cartridge_bootstrap_vshard: true
    cartridge_failover_params:
      mode: stateful
      state_provider: etcd2
      etcd2_params:
        prefix: some/prefix
        lock_delay: 30
        endpoints:
          - "http://192.168.16.11:2379"
          - "http://192.168.16.12:2379"
          - "http://192.168.16.11:2379"
  hosts:
    router-1:
      config:
        advertise_uri: "192.168.16.11:3031"
        http_port: 8081
    storage-1-1:
      config:
        advertise_uri: "192.168.16.11:3032"
        http_port: 8082
        METRICS_PATH: /health/metrics
        PROMETHEUS_PORT: "9999"
    storage-2-1:
      config:
        advertise_uri: "192.168.16.11:3033"
        http_port: 8083
        METRICS_PATH: /health/metrics
        PROMETHEUS_PORT: "9999"
    storage-3-1:
      config:
        advertise_uri: "192.168.16.11:3034"
        http_port: 8084
        METRICS_PATH: /health/metrics
        PROMETHEUS_PORT: "9999"
    storage-4-1:
      config:
        advertise_uri: "192.168.16.11:3035"
        http_port: 8085
        METRICS_PATH: /health/metrics
        PROMETHEUS_PORT: "9999"
    router-2:
      config:
        advertise_uri: "192.168.16.12:3031"
        http_port: 8081
    storage-1-2:
      config:
        advertise_uri: "192.168.16.12:3032"
        http_port: 8082
        METRICS_PATH: /health/metrics
        PROMETHEUS_PORT: "9999"
    storage-2-2:
      config:
        advertise_uri: "192.168.16.12:3033"
        http_port: 8083
        METRICS_PATH: /health/metrics
        PROMETHEUS_PORT: "9999"
    storage-3-2:
      config:
        advertise_uri: "192.168.16.12:3034"
        http_port: 8084
        METRICS_PATH: /health/metrics
        PROMETHEUS_PORT: "9999"
    storage-4-2:
      config:
        advertise_uri: "192.168.16.12:3035"
        http_port: 8085
        METRICS_PATH: /health/metrics
        PROMETHEUS_PORT: "9999"
  children:
    router-1-replicaset:
      vars:
        replicaset_alias: router-1
        failover_priority:
          - router-1
        roles:
          - router
          - failover-coordinator
      hosts:
        router-1: ~
    storage-1-replicaset:
      vars:
        replicaset_alias: storage-1
        failover_priority:
          - storage-1-1
          - storage-1-2
        roles:
          - storage
      hosts:
        storage-1-1: ~
        storage-1-2: ~
    storage-2-replicaset:
      vars:
        replicaset_alias: storage-2
        failover_priority:
          - storage-2-1
          - storage-2-2
        roles:
          - storage
      hosts:
        storage-2-1: ~
        storage-2-2: ~
    storage-3-replicaset:
      vars:
        replicaset_alias: storage-3
        failover_priority:
          - storage-3-1
          - storage-3-2
        roles:
          - storage
      hosts:
        storage-3-1: ~
        storage-3-2: ~
    storage-4-replicaset:
      vars:
        replicaset_alias: storage-4
        failover_priority:
          - storage-4-1
          - storage-4-2
        roles:
          - storage
      hosts:
        storage-4-1: ~
        storage-4-2: ~
    router-2-replicaset:
      vars:
        replicaset_alias: router-2
        failover_priority:
          - router-2
        roles:
          - router
          - failover-coordinator
      hosts:
        router-2: ~
    server-1:
      vars:
        ansible_host: 192.168.16.11
      hosts:
        router-1: ~
        storage-1-1: ~
        storage-2-1: ~
        storage-3-1: ~
        storage-4-1: ~
    server-2:
      vars:
        ansible_host: 192.168.16.12
      hosts:
        router-2: ~
        storage-1-2: ~
        storage-2-2: ~
        storage-3-2: ~
        storage-4-2: ~

